### PR TITLE
Luis omega/update odc

### PR DESCRIPTION
**Pull request by Luis-omega**
Closes # .

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/QueryM/DatumCacheWsp.purs [hunk 1]
```diff
@@ -104,24 +106,56 @@ instance DecodeAeson GetDatumsByHashesR where
...
+        obj
+
+      decodeValueOption
+        :: Aeson -> Either JsonDecodeError (Either String Datum)
+      decodeValueOption aes =
+        do
+          options <- (pure <$> getNestedAeson aes [ "value", "Right" ])
+            <|> (Left <$> getNestedAeson aes [ "value", "Left" ])
+          case options of
+            Left x -> pure $ Left $ show x
```
Comment (Luis-omega) [inline src/QueryM/DatumCacheWsp.purs:141]: This is the place to reproduce the possible error responses from ODC if we want to use a type instead of a String.

File: src/QueryM/DatumCacheWsp.purs [hunk 2]
```diff
@@ -104,24 +106,56 @@ instance DecodeAeson GetDatumsByHashesR where
...
+              $ identity <<< traverse decodeEmptyDatum
+          ) obj
+
+      decodeEmptyDatum
+        :: Aeson -> Either JsonDecodeError Unit
+      decodeEmptyDatum aes =
+        if stringifyAeson aes == "{}" then
+          pure unit
+        else
+          Left $ TypeMismatch "expected empty JSON"
```
Comment (jy14898) [inline src/QueryM/DatumCacheWsp.purs:126]:
```suggestion
      decodeEmptyDatum aes =
        unless (caseAesonObject false isEmpty aes) (Left $ TypeMismatch "Empty Object")
```

File: src/QueryM.purs [hunk 3]
```diff
@@ -467,7 +468,18 @@ getDatumByHash hash = unwrap <$> do
...
+    ( unwrap <$> do
+        mkDatumCacheRequest
+          DcWsp.getDatumsByHashesCall
+          _.getDatumsByHashes
+          hashes
+    )
+
+getDatumsByHashes'
+  :: Array DataHash -> QueryM (Map DataHash (Either String Datum))
+getDatumsByHashes' hashes = unwrap <$> do
```
[the remark this answers is not in the document]
Comment (Luis-omega) [reply src/QueryM.purs:482]: I did this on purpose since usually when I do things like 

```
someFunction = 
  do 
     action
```

which I prefer, I got corrected in the review to 

```
someFunction = do 
     action
```

As that's the style in the code base.

File: src/QueryM/DatumCacheWsp.purs [hunk 4]
```diff
@@ -104,24 +106,56 @@ instance DecodeAeson GetDatumsByHashesR where
...
+        ( decodeEmptyDatums =<< getNestedAeson
+            r
+            [ "DatumsFound", "value" ]
+        ) $> GetDatumsByHashesR Map.empty
+
       datumsNotFound =
         getNestedAeson r [ "DatumsNotFound" ] $> GetDatumsByHashesR Map.empty
     in
-      datumsFound <|> datumsNotFound
+      datumsFound <|> datumsEmpty <|> datumsNotFound
```
[the remark this answers is not in the document]
Comment (Luis-omega) [reply src/QueryM/DatumCacheWsp.purs:158]: There are two cases, the first is `The request of datums was made with an empty list of Hashes`, and the second is `No one of the datums was found`.

File: src/QueryM/DatumCacheWsp.purs (hunk 2 above)
Comment (Luis-omega) [reply src/QueryM/DatumCacheWsp.purs:126]: yes, I'm looking for an `isEmpty` function like this, but It doesn't seem to exist in the current Aeson api.

File: src/QueryM.purs [hunk 5]
```diff
@@ -467,7 +468,18 @@ getDatumByHash hash = unwrap <$> do
...
 getDatumsByHashes :: Array DataHash -> QueryM (Map DataHash Datum)
-getDatumsByHashes hashes = unwrap <$> do
+getDatumsByHashes hashes =
+  Map.mapMaybe hush <$>
+    ( unwrap <$> do
+        mkDatumCacheRequest
+          DcWsp.getDatumsByHashesCall
+          _.getDatumsByHashes
+          hashes
+    )
```
[the remark this answers is not in the document]
Comment (Luis-omega) [reply src/QueryM.purs]: We add the `'` suffix to avoid changing the API of `getDatumsByHashes`, it was suggested when in the discussion of the ODC update.

File: src/QueryM/DatumCacheWsp.purs (hunk 2 above)
Comment (jy14898) [reply src/QueryM/DatumCacheWsp.purs:122]: It's in Foreign.Object, which I believe is what the Object type here is (can't check ATM)

Comment (jy14898) [reply src/QueryM/DatumCacheWsp.purs:122]: https://pursuit.purescript.org/packages/purescript-foreign-object/4.0.0/docs/Foreign.Object#v:isEmpty

File: src/Contract/PlutusData.purs [hunk 6]
```diff
@@ -106,8 +109,17 @@ getDatumByHash
...
+-- | This function discards all possible error getting a `DataHash`.
 getDatumsByHashes
   :: forall (r :: Row Type)
    . Array DataHash
   -> Contract r (Map DataHash Datum.Datum)
 getDatumsByHashes = wrapContract <<< QueryM.getDatumsByHashes
+
+-- | Get `PlutusData`s given a an `Array` of `DataHash`.
+-- | In case of error, the returned string contains the needed information.
+getDatumsByHashes'
```
Comment (jy14898) [inline src/Contract/PlutusData.purs:121]: For completeness, lets also replicate the test in `Test.AffInterface.testOgmiosDatumCacheGetDatumsByHashes` for this call

Comment (jy14898) [reply src/Contract/PlutusData.purs:121]: And `Test.Plutip`'s `runPlutipContract: Datums`

File: src/QueryM/DatumCacheWsp.purs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jy14898) [reply src/QueryM/DatumCacheWsp.purs:154]: If you've made a request with an empty array of hashes, won't `datumsFound` succeed? Because it traverses the response array, which because it's empty will run no effects and thus succeed. Maybe it would help me to understand if I saw example requests + responses

[the remark this answers is not in the document]
Comment (Luis-omega) [reply src/QueryM/DatumCacheWsp.purs:153]: When I was writing the ODC update I was trying to avoid API changes, but now it seems that decision wasn't good in the end. As a result, we have some complexity here, I would make an issue for a change so that we would see the following responses.

With the new proposal we would get in case of an empty list : 

`{"DatumsFound" : {"value": [  ] } } `

And if we can't find some datums : 

`{"DatumsFound" : {"value": [ {"hash":"SomeHash", "value":{"Left":"DatumNotFound"}} ... ] } } `
 


The current code in case of an empty list of datums returns :  

`{"DatumsFound" : {"value": [ {} ] } } `

And If we can't find any of the datums also returns : 

`{"DatumsFound" : {"value": [ {} ] } } `

[the remark this answers is not in the document]
Comment (Luis-omega) [reply src/QueryM/DatumCacheWsp.purs:153]: the issues is here https://github.com/mlabs-haskell/ogmios-datum-cache/issues/125

[the remark this answers is not in the document]
Comment (Luis-omega) [reply src/QueryM/DatumCacheWsp.purs]: The correction is waiting for merge but this pr is now using it.

File: src/Contract/PlutusData.purs (hunk 6 above)
Comment (Luis-omega) [reply src/Contract/PlutusData.purs:125]: I going to add `Test.AffInterface.testOgmiosDatumCacheGetDatumsByHashes'` but those checks are more or less pointless right now since the `hash` used in the tests is outdated and all the test returns the same edge case of not datums founds.

File: src/QueryM/DatumCacheWsp.purs (hunk 2 above)
Comment (Luis-omega) [reply src/QueryM/DatumCacheWsp.purs]: Thanks! I didn't know I can use it like this. Sadly this isn't needed anymore with the last change in ODC api.

File: src/Contract/PlutusData.purs [hunk 7]
```diff
@@ -109,9 +112,18 @@ getDatumByHash
...
+-- | This function discards all possible error getting a `DataHash`.
 getDatumsByHashes
   :: forall (r :: Row Type)
    . Array DataHash
   -> Contract r (Map DataHash Datum.Datum)
 getDatumsByHashes = wrapContract <<< QueryM.getDatumsByHashes
+
+-- | Get `PlutusData`s given an `Array` of `DataHash`.
+-- | In case of error, the returned string contains the needed information.
+getDatumsByHashes'
```
Comment (klntsky) [inline src/Contract/PlutusData.purs:125]: Let's use descriptive names rather than `'`.

```suggestion
getDatumsByHashesWithErrors
```

**Review by klntsky — CHANGES_REQUESTED** [review]

**Review by jy14898 — APPROVED** [review]
LGTM, I'm not sure if we should hold merging until the changes are merged in to ODC too

**Comment by klntsky** [comment]
@jy14898 we should

File: CHANGELOG.md [hunk 8]
```diff
@@ -136,6 +136,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - KeyWallet's `selectCollateral` field now allows multiple collateral to be selected, and is provided with `coinsPerUtxoByte` and `maxCollateralInputs` from the protocol parameters. ([#947](https://github.com/Plutonomicon/cardano-transaction-lib/pull/947))
 - `mustPayWithDatumToPubKey`, `mustPayWithDatumToPubKeyAddress`, and `mustPayToScript` now expect a `DatumPresence` tag in their arguments to mark whether the datum should be inline or hashed in the transaction output. ((#931)[https://github.com/Plutonomicon/cardano-transaction-lib/pull/931])
 - Switched to [blakejs](https://github.com/dcposch/blakejs) for blake2b hashing. `blake2b256Hash` and `blake2b256HashHex` functions are now pure ([#991](https://github.com/Plutonomicon/cardano-transaction-lib/pull/991))
+- Updated ODC version, this includes a new function `getDatumsByHashesWithErrors:: Contract r (Map DataHash (Either String Datum))`, it returns a `String` whenever ODC have an error retrieving the corresponding `DataHash`. This update also changes the way we store transactions in the local database, meaning that we need to drop the `transactions` table.
```
Comment (klntsky) [inline CHANGELOG.md:139]: The type is not correct. Please remove the type signature.

```suggestion
- Updated ODC version, this includes a new function `getDatumsByHashesWithErrors` that does not discard errors, unlike `getDatumsByHashes`. **This update also changes the way we store transactions in the local database, meaning that we need to drop the `transactions` table.**
```

**Review by klntsky — CHANGES_REQUESTED** [review]

**Comment by Luis-omega** [comment]
The changes are already merged in ODC, I have already updated the Hash of ODC.

File: CHANGELOG.md [hunk 9]
```diff
@@ -143,6 +143,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - KeyWallet's `selectCollateral` field now allows multiple collateral to be selected, and is provided with `coinsPerUtxoByte` and `maxCollateralInputs` from the protocol parameters. ([#947](https://github.com/Plutonomicon/cardano-transaction-lib/pull/947))
 - `mustPayWithDatumToPubKey`, `mustPayWithDatumToPubKeyAddress`, and `mustPayToScript` now expect a `DatumPresence` tag in their arguments to mark whether the datum should be inline or hashed in the transaction output. ((#931)[https://github.com/Plutonomicon/cardano-transaction-lib/pull/931])
 - Switched to [blakejs](https://github.com/dcposch/blakejs) for blake2b hashing. `blake2b256Hash` and `blake2b256HashHex` functions are now pure ([#991](https://github.com/Plutonomicon/cardano-transaction-lib/pull/991))
+- Updated ODC version, this includes a new function `getDatumsByHashes`that does not discard errors, unlike `getDatumsByHashes`. **This update also changes the way we store transactions in the local database, meaning that we need to drop the `transactions` table.**
```
Comment (klntsky) [inline CHANGELOG.md:146]:
```suggestion
- Updated ODC version, this includes a new function `getDatumsByHashesWithErrors` that does not discard errors, unlike `getDatumsByHashes`. **This update also changes the way we store transactions in the local database, meaning that we need to drop the `transactions` table.**
```

**Review by klntsky — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -143,6 +143,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - KeyWallet's `selectCollateral` field now allows multiple collateral to be selected, and is provided with `coinsPerUtxoByte` and `maxCollateralInputs` from the protocol parameters. ([#947](https://github.com/Plutonomicon/cardano-transaction-lib/pull/947))
 - `mustPayWithDatumToPubKey`, `mustPayWithDatumToPubKeyAddress`, and `mustPayToScript` now expect a `DatumPresence` tag in their arguments to mark whether the datum should be inline or hashed in the transaction output. ((#931)[https://github.com/Plutonomicon/cardano-transaction-lib/pull/931])
 - Switched to [blakejs](https://github.com/dcposch/blakejs) for blake2b hashing. `blake2b256Hash` and `blake2b256HashHex` functions are now pure ([#991](https://github.com/Plutonomicon/cardano-transaction-lib/pull/991))
+- Updated ODC version, this includes a new function `getDatumsByHashesWithErrors`that does not discard errors, unlike `getDatumsByHashes`. **This update also changes the way we store transactions in the local database, meaning that we need to drop the `transactions` table.**
```
Comment (ngua) [inline CHANGELOG.md:146]: Thanks for adding the reminder! ODC changes have been an issue in the past, it's nice to have the information in one place

**Review by ngua — APPROVED** [review]
LGTM. Since the ODC PR has been merged, I think it's OK to merge this
